#2838 standardize nullability annotations

AlexanderSkrock opened the pull request:
- Fixes #2838

knutwannheden commented on rewrite-gradle/src/main/resources/META-INF/rewrite/nullability.yml:
(Trying out GitHub file comments for the first time :-))

I think I undestand that you add this declarative recipe (and the other one for Gradle) so that the corresponding dependency automatically gets added. I understand the reasoning behind this, but personally I would be inclined to wait with this. There is an ongoing effort to provide a recipe which will allow adding a dependency regardless of which build system is used (as long as OpenRewrite supports it). So if we can wait for that feature, I would prefer if we do that.

Another concern is that I don't know how many of the users want to migrate to `javax.annotation.Nullable` (especially with the Jakarta annotations being available). The user may also want to migrate to JSpecify. This is another reason I think we should at least wait with this declarative recipe, which also adds the dependency.

As an alternative we maybe find a solution to register a recipe in Java code which adds the correct dependency.

AlexanderSkrock replied:
> I think I undestand that you add this declarative recipe (and the other one for Gradle) so that the corresponding dependency automatically gets added. I understand the reasoning behind this, but personally I would be inclined to wait with this. There is an ongoing effort to provide a recipe which will allow adding a dependency regardless of which build system is used (as long as OpenRewrite supports it). So if we can wait for that feature, I would prefer if we do that.

That is basically what I thought while implementing. First I had those in one recipe within `rewrite-java`, but then I realized how the modules are connected to each other, that `rewrite-maven` and `rewrite-gradle` depend on `rewrite-java`. Splitting these recipes up did not give me a good feeling.
I'm glad that you already picked up the idea of being independent of the dependency management, let it be Gradle, Maven or something completely different. I'm glad to hear you are working on this and would also volunteer to integrate it once implemented.

But then, I would rather keep dependencies out of scope instead of manually implementing something. That is also how other recipe currently handle this. However decides to go for annotation `X` or annotation `Y` is probably aware of the dependency to be added anyways or it is already defined as dependency.

knutwannheden commented on rewrite-gradle/src/test/java/org/openrewrite/gradle/java/UseJavaxNullabilityAnnotationsTest.java:
```diff
@@ -0,0 +1,121 @@
+ * <br>
+ *
+ * @author askrock
+ */
```
I suggest we remove the Javadoc above.

AlexanderSkrock replied:
I did not mean to commit that, that's how it was configured on my other device, sorry about that. 🙈

knutwannheden commented on rewrite-java/src/main/java/org/openrewrite/java/StandardizeNullabilityAnnotations.java:
```diff
@@ -0,0 +1,99 @@
+            "org.jmlspecs.annotation.Nullable",
+            "org.openrewrite.internal.lang.Nullable",
+            "lombok.NonNull"
+    ).collect(Collectors.toSet());
```
Neat trick when no `Set.of()` is available.

knutwannheden commented on rewrite-java/src/main/java/org/openrewrite/java/StandardizeNullabilityAnnotations.java:
```diff
@@ -0,0 +1,99 @@
+            "org.springframework.lang.Nullable",
+            "org.jmlspecs.annotation.Nullable",
+            "org.openrewrite.internal.lang.Nullable",
+            "lombok.NonNull"
```
This last one looks like it belongs in the other set. Here is a nice list of other similar annotations: https://checkerframework.org/manual/#nullness-related-work. I don't think we need all of these, but the JSpecify annotations would be nice to add, as that framework has gained quite a bit of traction.

AlexanderSkrock replied:
That's another good source. Thank you for pointing out! I will have a look on them in detail.

knutwannheden commented on rewrite-maven/src/main/resources/META-INF/rewrite/nullability.yml:
As above, I suggest we wait with this part.

knutwannheden commented on rewrite-java/src/main/java/org/openrewrite/java/StandardizeNullabilityAnnotations.java:
```diff
@@ -0,0 +1,99 @@
+
+
+@Value
+@NonNullApi
```
We already have this in the `package-info.java` file, so we can remove it here.

knutwannheden commented on rewrite-maven/src/test/java/org/openrewrite/maven/java/UseJavaxNullabilityAnnotationsTest.java:
```diff
@@ -0,0 +1,103 @@
+ * <br>
+ *
+ * @author askrock
+ */
```
Can we also remove this comment?

Please also note that your contribution will in addition to the release notes also be attributed in the Moderne SaaS (as extracted from the Git history), as you can see in the bottom of this screenshot :-) 
<img width="518" alt="image" src="https://user-images.githubusercontent.com/408765/230323193-e9fbd770-272d-4e18-8109-4852bec56670.png">

AlexanderSkrock replied:
I did not mean to commit that, that's how it was configured on my other device, sorry about that. 🙈

knutwannheden commented on rewrite-java-test/src/test/java/org/openrewrite/java/StandardizeNullabilityAnnotationsTest.java:
```diff
@@ -0,0 +1,274 @@
+        spec.recipe(new StandardizeNullabilityAnnotations(Nullable.class.getName(), NonNull.class.getName()));
+    }
+
+    @Test
```
Can we also add a test for a `package-info.java` file?

```java
@NonNullApi
@NonNullFields
package org.openrewrite.java;

import org.openrewrite.internal.lang.NonNullApi;
import org.openrewrite.internal.lang.NonNullFields;
```

AlexanderSkrock replied:
Good idea! Will also add one for class-level annotations.

knutwannheden reviewed (commented):
I will also ask @sambsnyd to weigh in with his view on this.

sambsnyd reviewed (approved):
Looks useful to me. Thanks!

AlexanderSkrock commented:
Between holidays, I had time to think about this recipe and committed a work in progress version. What do you think about this approach, @knutwannheden. In my opinion, we can solve the standardization issue even better this way. Would be more flexible and robust. Anyways, I you prefer the original approach, I will roll back to the last commit before the overhaul and simply add your proposed testcase and nullability annotation.

timtebeek commented on rewrite-core/src/main/java/org/openrewrite/internal/ListUtils.java:
```diff
@@ -276,4 +276,30 @@ public static <T> List<T> insertAll(@Nullable List<T> ls, int index, @Nullable L
+
+    /**
+     * Return the first element of the given list if available
+     * @return the last element or {@code null} if the list is empty
```
```suggestion
     * @return the first element or {@code null} if the list is empty
```

AlexanderSkrock commented:
Sadly, I'm only able to do small steps, as I barely have coherent time slots. The current state of my draft merge request is as follows:
- Replacement works on various elements
- Replacement works with different scopes
- Adding and removing import works

ToDos:
- A general problem when inserting annotations that type resolution of inserted annotations does not work. Testcases get the expected code but type validation assertion fails.
- For return type annotations the spacing is odd, as it misses a whitespace between the visibility modifier and the annotation. E. g. `public@NonNull`. I attempted to modify autoformatting, but I did not find a suitable solution yet, which works in all cases.
- Adding targets and scopes for all known nullability annotations

timtebeek commented:
> Sadly, I'm only able to do small steps, as I barely have coherent time slots.

We appreciate all the work you've put in already! No worries about timing; do let us know if you'd like to hand off some of the work, otherwise we'll wait for you as we don't want to take away the fun seeing a feature through.
 
> * A general problem when inserting annotations that type resolution of inserted annotations does not work. Testcases get the expected code but type validation assertion fails. That's my core problem that remains. I understand that type is unknown as the import is added `afterVisit`, so after the template was executed. Tho, I do not know a solution yet.

Hoping for instance @knutwannheden can chime in here; as he went through your PR before already, before I dive in as well.

> * For return type annotations the spacing is odd, as it misses a whitespace between the visibility modifier and the annotation. E. g. `public@NonNull`. I attempted to modify autoformatting, but I did not find a suitable solution yet, which works in all cases.

Typically in such cases you might be lacking a prefix, and I don't see a mention of prefix in `StandardizeNullabilityAnnotationsVisitor` so far. You can likely find other examples where we add the prefix of the original element on the replacement element, which might fix the issues you're having here as well.

knutwannheden commented:
> Hoping for instance @knutwannheden can chime in here; as he went through your PR before already, before I dive in as well.

@AlexanderSkrock Yes, please don't hesitate to reach out if you have questions or problems you don't know how to solve. We will do our best to help out. I will try to take a look at your PR. So far it looks like you have made a lot of good progress!

knutwannheden commented on rewrite-java/src/main/java/org/openrewrite/java/nullability/StandardizeNullabilityAnnotations.java:
```diff
@@ -0,0 +1,183 @@
+    @Option(displayName = "Nullability annotations to use",
+            description = "All other nullability annotations will be replaced by these.",
+            required = false)
+    Set<NullabilityAnnotation> additionalNullabilityAnnotations;
```
I think it would make sense to change the `displayName` and `description` slightly, so that it is clear what the difference is between `nullabilityAnnotationsFqn` and `additionalNullabilityAnnotations`.

AlexanderSkrock replied:
Oh, forgot to change these, thank you spotting!

knutwannheden reviewed (commented):
Coming along nicely. I will look at it more in depth soon.

AlexanderSkrock commented on rewrite-core/src/main/java/org/openrewrite/internal/ListUtils.java:
```diff
@@ -276,4 +276,30 @@ public static <T> List<T> insertAll(@Nullable List<T> ls, int index, @Nullable L
+     * @return the last element or {@code null} if the list is empty
+     */
+    @Nullable
+    public static <T> T last(@Nullable List<T> ls) {
```
After reading your comment, I actually wonder why my `ListUtils#first` method would be a benefit, except for the null check. :D
I agree with you that the possible `null` return value would reduce the usability again, so in the end plain java `myList.get(myList.size() - 1)` is probably the better approach for now.

knutwannheden commented on rewrite-java/src/main/java/org/openrewrite/java/nullability/StandardizeNullabilityAnnotations.java:
```diff
@@ -0,0 +1,183 @@
+    @JsonCreator
+    public StandardizeNullabilityAnnotations(List<String> nullabilityAnnotationsFqn, Set<NullabilityAnnotation> additionalNullabilityAnnotations) {
+        this.nullabilityAnnotationsFqn = nullabilityAnnotationsFqn;
+        this.additionalNullabilityAnnotations = ObjectUtils.defaultIfNull(additionalNullabilityAnnotations, new HashSet<>());
```
Right. I keep forgetting that we still target Java 8 😄

AlexanderSkrock commented:
> > Hoping for instance @knutwannheden can chime in here; as he went through your PR before already, before I dive in as well.
> 
> @AlexanderSkrock Yes, please don't hesitate to reach out if you have questions or problems you don't know how to solve. We will do our best to help out. I will try to take a look at your PR. So far it looks like you have made a lot of good progress!

Thank for your feedback already!

Except for completing the list of known nullability annotations, I am almost done. So far, one problem left when the replacing annotation has the same simple name.

Consider the following snippet:
```java
J.ClassDeclaration cleanClassDeclaration = classDeclaration.withLeadingAnnotations(cleanedAnnotations);
return annotationsForReplacement.stream().reduce(
        cleanClassDeclaration,
        (J.ClassDeclaration currentDeclaration, NullabilityAnnotation annotation) -> currentDeclaration.withTemplate(
                annotationTemplate(annotation),
                currentDeclaration.getCoordinates().addAnnotation(Comparator.comparing(J.Annotation::getSimpleName))
        ), (oldDeclaration, newDeclaration) -> newDeclaration
);
```
The annotations are already replaced, but adding and removing imports is delayed to `afterVisit`. Now, when we use the template to add the replacement annotations, we have the old imports and new imports (that we manually added to the template). If the old and the new annotation now have the same simple name, the parser can not resolve the annotation's type properly because both imports match.
Do you have an idea on this, @knutwannheden ? As far as I understood, I would not have the option at this point to modify imports directly while visiting e. g. the `ClassDeclaration`, right?

knutwannheden commented:
That is indeed a tricky problem at the moment. We intend to improve this by allowing a template to contain fully qualified references, which then optionally (and if possible without conflict) get "shortened" again with a corresponding import when updating the LST.

I have a few ideas how to handle this. You could try using `ChangeType`, but I don't know if it can handle this case either. Otherwise you could first remove the import and annotations (and remember the locations) and in a second step (using `doAfterVisit()` or by hooking it in inside a `visitSourceFile()` override) add the new annotations (and import). Alternatively, but this should be last resort, directly attempt to update the type annotations. I am saying last resort, because this is often rather tricky to get exactly right.

Please let me know how it goes and if you don't find a good solution I will try to dig deeper.

AlexanderSkrock commented:
First of all, thank you for your ideas!

I like that two-staged approach by splitting removal of old annotations and insertion of new annotations. I almost started implementing, when I realized there are really odd edge-cases which may require me to possibly use fully-qualified identifiers anyways if necessary, which is more complex.

### Case 1
Suppose following:
- Library A has annotation `NonNullFields` and `NonNullApi`
- Library B has annotation `NonNullFields`

Within one source file we now have two locations:
```java
@NonNullFields
public static class X {
}
```
and
```java
@NonNullApi
@NonNullFields
public static class Y {
}
```
On class `X` we can easily replace A's `@NonNullFields` with B's `@NonNullFields`.
On class `Y` we will realize we can not replace both annotations, because our library `B` only supports `@NonNullFields`. As consequence, we do not touch class `Y`.
=> In the end we kept importing A's `@NonNullFields`, but also added B's `@NonNullFields` which now clash.

### Case 2
Suppose following:
- Library A has annotation `NonNull` which does not work on packages (e. g. `org.openrewrite.internal.lang.NonNull`)
- Library B has annotation `NonNull` which does not work on packages and classes (e. g. `org.springframework.lang.NonNull`)

Within our code base we have the following class:
```java
@NonNull
public static class Y {

    @NonNull
    String nonNullVariable = "";
}
```

On the member variable we can easily replace A's `@NonNull` with B's `@NonNull`
On the class we might not have a replacement (for the example spring-core, there would be the option to replace with `@NonNullFields` and `@NonNullApi`, but let us forget that for a second). As consequence we do not touch the annotation on the class.
=> In the end we kept importing A's `@NonNull`, but also added B's `@NonNull` which now clash.

knutwannheden commented:
Yes, your use cases sound valid, unfortunately. And there are probably also other recipes which could have similar issues, which is why we are thinking about solving it in `JavaTemplate`.

Maybe we could for the time being do it this way then: Replace the annotations with fully qualified annotations, remove the import and finally have a visitor which adds an import (assuming there is no more conflict) and replaces the fully qualified references. This latter step was recently requested as a recipe anyway and would be fairly easy to achieve even without JavaTemplate.

So you don't have to solve yet another problem, I could look into providing this last recipe. Unfortunately I am on vacation next week. So unless you feel inclined to tackle that as well (in a separate PR) or find some other solution, you might have to wait another week, unless I find an hour or so to look into this.

AlexanderSkrock commented:
Sounds promising and I would be glad if you could implement that functionality. Anyway, if you feel having too few time, I could also give it a try. Either way, no hurry on that. Meanwhile I will finish the collection of known nullability annotation.

AlexanderSkrock commented:
@knutwannheden , the recipe you thought of is the same as in this issue it seems: https://github.com/openrewrite/rewrite/issues/3094

knutwannheden commented:
Correct, that's the one.

knutwannheden commented:
@AlexanderSkrock #3094 is now fixed and you should be able to use `doNext()` to shorten the fully qualified type references.

AlexanderSkrock commented:
Thank you a lot, @knutwannheden ! I will try to integrate that recipe asap.

AlexanderSkrock commented:
At first glance, there are two issues with `ShortenFullyQualifiedTypeReferences`:
- It ignore types that are available within the package
- It shortens by simple name and thereby might shorten types with same simple name but different fully qualified name

knutwannheden commented:
I fixed a couple of more edge cases around the shortening of qualified names, including what you mentioned regarding types in the same package.

AlexanderSkrock commented:
Almost done! Seems only logging is missing to notify the user, whenever replacement was not possible, so these occurences can possibly do manual steps to migrate. Tho, I would keep these on debug level, to not spam on casual runs.

I saw that `slf4j` is used throughout the project. Does OpenRewrite also come if an specific implementation of that api, that I should add as dependency to `rewrite-java`? I was not able to spot one. Or doesn't OpenRewrite come with a specific implementation and it is up to the user of OpenRewrite?

AlexanderSkrock commented on rewrite-java/src/main/java/org/openrewrite/java/nullability/StandardizeNullabilityAnnotationsVisitor.java:
```diff
@@ -0,0 +1,326 @@
+        if (matchedNullabilityAnnotations == null || matchedNullabilityAnnotations.isEmpty()) {
+            return jPackage;
+        }
+        List<NullabilityAnnotation> annotationsForReplacement = getAnnotationsForReplacement(matchedNullabilityAnnotations, Stream.of(NullabilityAnnotation.Target.PACKAGE).collect(Collectors.toSet()));
```
Honestly, I knew `EnumMap`, but I did not know about a similar set implementation. This simplifies many calls troughout my code, especially as we can not use `Set#of`!

AlexanderSkrock commented on rewrite-java/src/main/java/org/openrewrite/java/nullability/StandardizeNullabilityAnnotationsVisitor.java:
```diff
@@ -0,0 +1,326 @@
+            // Maybe we should transfer the logging to the caller
+            // as that code knows more about the element we failed at.
+        }
+        if (annotationsForReplacement.containsAll(usedNullabilityAnnotations) && usedNullabilityAnnotations.containsAll(annotationsForReplacement)) {
```
You are right, it should be more efficient to compare sets instead of this workaround. I will change as you proposed, because I would like to maintain order. This way we replace annotations in a deterministic manner. Another advantage ist, that you can define the "priority" for replacement inside configuration. E. g. if you do not use a library which provides all nullability annotations you need, you could also configure the recipe to use a combination of ``Jetbrains` and a second library which has an annotation that also applies on package-level while ensuring that we use `Jetbrains` where possible.

AlexanderSkrock commented on rewrite-java/src/main/java/org/openrewrite/java/nullability/StandardizeNullabilityAnnotationsVisitor.java:
```diff
@@ -0,0 +1,326 @@
+                        annotationTemplate(annotation),
+                        currentType.getCoordinates().addAnnotation(Comparator.comparing(J.Annotation::getSimpleName))
+                ), (oldType, newType) -> newType
+        );
```
I assume you would like to insert the annotations at the same position where we removed them from. Basically we could do this using `List#indexOf` to retrieve the previous indices and later do the same inside the comparator for insertion.

We could remember the first index we removed an annotation from:
`int firstIndexWeRemoveAnnotations = jPackage.getAnnotations().indexOf(currentNullabilityAnnotations.get(0));`
Comparator respecting the remembered insertion point:
```java
(a1, a2) -> {
    int previousIndexA1 = jPackage.getAnnotations().indexOf(a1);
    int previousIndexA2 = jPackage.getAnnotations().indexOf(a2);
    if (previousIndexA1 < 0 && previousIndexA2 < 0) {
        return Comparator.comparing(J.Annotation::getSimpleName).compare(a1, a2);
    } else if (previousIndexA1 < 0) {
        return previousIndexA2 >= firstIndexWeRemoveAnnotations ? -1 : 1;
    } else if (previousIndexA2 < 0) {
        return previousIndexA1 < firstIndexWeRemoveAnnotations ? -1 : 1;
    } else {
        return 0;
    }
}
```
But this would always insert at the first position. In cases where we replace e. g. two annotations with two annotations would insert both new annotations at the position of the first that we removed.

I understand the reasoning behind this, but I am not sure how "perfect" we are able to achieve that. Also, the code snippet is quite long for such a "trivial purpose", but there's no api yet to insert at specific indices.

knutwannheden replied:
Yes, it doesn't appear to be that easy at the moment, unfortunately. Let's keep it like this for now then. Thanks for clarifying!

timtebeek commented on rewrite-java/src/main/resources/META-INF/rewrite/nullability.yml:
```diff
@@ -0,0 +1,68 @@
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.nullability.UseJakartaNullabilityAnnotations
+displayName: Use jakarte nullability annotations
```
```suggestion
displayName: Use Jakarta nullability annotations
```

timtebeek commented on rewrite-java/src/main/resources/META-INF/rewrite/nullability.yml:
```diff
@@ -0,0 +1,68 @@
+        - jakarta.annotation.Nonnull
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.nullability.UseSpringNullabilityAnnotations
```
I think we can keep this here, rather than spread this recipe use across multiple repositories. There's little risk keeping this here I suppose; the spring specifics are only in configuration values and don't require a new runtime dependency.

timtebeek commented on rewrite-java/src/main/resources/META-INF/rewrite/nullability.yml:
```diff
@@ -0,0 +1,68 @@
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.nullability.UseSpringNullabilityAnnotations
+displayName: Use spring nullability annotations
+description: Replaces other known nullability annotations by spring annotations where possible.
```
Generally I'd use "replace with" rather than "replace by" in all our descriptions, but English is not my mother tongue, so I  might be wrong

AlexanderSkrock replied:
I think these words are interchangeable, but also, english is not my mother tongue. I am fine going with "with" :D

knutwannheden reviewed (approved):
Apart from the last remark by Tim I think this PR is good to go.

AlexanderSkrock commented:
> There is actually one case which isn't addressed yet. The Java language for some reason allows annotations to appear in-between modifiers. So it is actually allowed to have a class like this:
> 
> ```java
>             class Test {
>                 private
>                 @Nonnull
>                 final
>                 String variable = "";
>             }
> ```
> 
> In OpenRewrite the LST is also a bit unusual in this case. There is a `J.Modifier` type which declares an `annotations` field. I think it should probably be straight-forward to extend the recipe to also support this corner case, or what do you think @AlexanderSkrock ?

Oh, I overlooked that one! Yeah, I could implement it within `visitClassDeclaration`, but for annotations that belong to modifiers `visitAnnotation` would not be called, am I right? In consequence we could only collect annotations manually from class declaration. Just writing that, I realize the code currently does not check whether an annotation was found within "annotations" or "leading annotations".

knutwannheden commented:
> Oh, I overlooked that one! Yeah, I could implement it within `visitClassDeclaration`, but for annotations that belong to modifiers `visitAnnotation` would not be called, am I right? I came to this conclusion because `J.Modifiers` are not visisted with their specialized type, but as casual statements. (I will verify this with a testcase quickly.) In consequence we could only collect annotations manually from class declaration. Writing that, I realize the code currently does not check whether an annotation was found within "annotations on a classes kind" or "leading annotations". This adds another question, similary to what you asked earlier about "insertion point of annotations", should we always insert at leading annotations? That's what we currently do (and probably is the most used option), still it is possible to do it between modifiers and after modifiers.

For `J.Modifier` you could add a `visitModifier()` overload to your visitor to collect data about these annotations. Please also note that all of `J.ClassDeclaration`, `J.MethodDeclaration`, and `J.VariableDeclarations` declare an `getAllAnnotations()` method, which collects all annotations from the places they can appear in.

> Last, but not least, for the first and second bullet, I would propose we define inserting annotations in front (e. g. for `ClassDeclaration` as leading annotation) as standard.

In the JLS there is [this remark](https://docs.oracle.com/javase/specs/jls/se20/html/jls-9.html#jls-9.7.4):
> It is customary, though not required, to write declaration annotations before all other modifiers, and type annotations immediately before the type to which they apply.

So, either we could try to replace the annotations exactly in the same place as they were declared (but as previously discussed, this is currently not as easy as one would like it to be) or we could instead just always add the new annotations as leading annotations (all of the previously mentioned types also declare a `leadingAnnotations` field), which would be in line with the remark from the JLS and your proposal.

sambsnyd commented:
If the recipe moves the strangely-placed annotations in-place that is fine, this recipe's purpose will have been accomplished.
If the recipe moves the annotation to the usual/canonical place for annotations that is a bonus. 
Regarding either result as acceptable, I would probably go with whichever is easier/clearer to implement.

knutwannheden commented:
@AlexanderSkrock Thanks for the heads-up and I wish you and your family all the best 🍀  Since the PR is nearly done, I think we will try to finish it. Thank you again for all the work you put into it!

timtebeek commented:
With JSpecify coming out we've done a write up of our migration from internal annotations to JSpecify:
https://www.moderne.ai/blog/mass-migration-of-nullability-annotations-to-jspecify
The lessons learned and recipes developed there are likely to be valuable in this effort as well.

timtebeek commented:
FYI: There's now a parallel effort in
- https://github.com/openrewrite/rewrite-migrate-java/pull/541

timtebeek commented:
With the release of JSpecify we've created separate recipes in rewrite-migrate-java to migrate from various other libraries to JSpecify: https://docs.openrewrite.org/recipes/java/jspecify/migratetojspecify

I propose we close this effort, and pull over anything that's still applicable to rewrite-migrate-java instead. Thanks again @AlexanderSkrock !
